lib/spack/spack/util/gpg.py

```diff
@@ -83,64 +348,105 @@ def create(cls, **kwargs):
-        args.extend(keys)
-        cls.gpg()(*args)
-
-    @classmethod
-    def sign(cls, key, file, output, clearsign=False):
-        args = [
-            '--armor',
-            '--default-key', key,
-            '--output', output,
-            file,
-        ]
-        if clearsign:
-            args.insert(0, '--clearsign')
-        else:
-            args.insert(0, '--detach-sign')
-        cls.gpg()(*args)
+            skeys = self.signing_keys(*keys)
+            self('--batch', '--yes', '--delete-secret-keys', *skeys)
 
-    @classmethod
-    def verify(cls, signature, file, suppress_warnings=False):
-        if suppress_warnings:
-            cls.gpg()('--verify', signature, file, error=str)
-        else:
-            cls.gpg()('--verify', signature, file)
+        pkeys = self.public_keys(*keys)
+        self('--batch', '--yes', '--delete-keys', *pkeys)
 
-    @classmethod
-    def list(cls, trusted, signing):
+    @wrap
+    def sign(self, key, file, output, clearsign=False):
+        self(('--clearsign' if clearsign else '--detach-sign'),
+             '--armor', '--default-key', key,
+             '--output', output, file)
+
+    @wrap
+    def verify(self, signature, file, suppress_warnings=False):
+        self('--verify', signature, file,
+             **({'error': str} if suppress_warnings else {}))
+
+    @wrap
+    def list(self, trusted, signing):
         if trusted:
-            cls.gpg()('--list-public-keys')
+            self('--list-public-keys')
+
         if signing:
-            cls.gpg()('--list-secret-keys')
+            self('--list-secret-keys')
 
 
 class SpackGPGError(spack.error.SpackError):
     """Class raised when GPG errors are detected."""
+
+
+# Make wrapped versions of the Gpg instance methods for convenience
+#
+# Done so that most calling code can skip the creation of a Gpg instance.
+#
+#     Instead of this:
+#         import spack.util.gpg
+#         gpg = spack.util.gpg.Gpg(gnupg_home='...')
+#         gpg('--version')
+#         gpg.public_keys()
+#
+#     Use this:
+#         import spack.util.gpg as gpg
+#         gpg.gpg('--version')
+#         gpg.public_keys()
+#
+#     If using a non-default GNUPGHOME:
+#         with gpg.gnupg_home_override('...'):
+#             ...
+def _make_wrapped_callables(namespace):
+    global wrap
+
+    def _make_wrapped_callable(func, name):
+        if func.__name__ == '__call__':
+            @functools.wraps(func)
+            def result(*args, **kwargs):
+                _callable = get_global_gpg_instance()
+                return _callable(*args, **kwargs)
+        else:
+            @functools.wraps(func)
+            def result(*args, **kwargs):
+                _callable = getattr(get_global_gpg_instance(), name)
+                return _callable(*args, **kwargs)
+
+        result.name = name
+        return result
+
+    for func, name in wrap:
+        namespace[name] = _make_wrapped_callable(func, name)
+
+
+_make_wrapped_callables(globals())
```

Comment:
My $0.02: I'm not a fan of the meta-programming approach taken here, seems like a lot of complexity for the benefit of users not having to create an instance of the `Gpg` class.  It took a non-zero amount of time and effort to figure out what you were doing with the various bits: the mysterious and uncommented `wrap` decorator class that inherits from `list`, its global instance used in decorating selected methods of the `Gpg` class, the definition and subsequent deletion of this method where you generate the module level handles from stuff stored by the decorator, etc...

Again, just my opinion, but I found all that stuff kind of a pain to review, and I would have preferred something straightforward like explicit definition of the handful of module-level methods you want to expose, in terms of a singleton `Gpg` instance.  Maybe I'm just exposing myself as simple-minded here though.

Reply:
Not at all!  I hate it, too (and I'm the one who wrote it!).

I think there's a neat meta programming idea here that could be pulled out and stashed away in some module somewhere, but frankly I was just going for future-proofing, and sometimes I just need someone to point out when I'm being too clever for my own good!

I agree this should be simpler.  I'll take another pass and see how it turns out.

Reply:
Have a look at the latest commit.  Is that easier to make sense of?